esdt-safe/src/to_sovereign/create_tx.rs

```diff
@@ -112,114 +115,205 @@ pub trait CreateTxModule:
+            {
+                refundable_payments.push(payment.clone());
+
+                continue;
+            } else {
                 total_tokens_for_fees += 1;
             }
 
-            if payment.token_nonce > 0 {
-                let current_token_data = self.blockchain().get_esdt_token_data(
-                    &own_sc_address,
+            let mut current_token_data = self.blockchain().get_esdt_token_data(
+                &own_sc_address,
+                &payment.token_identifier,
+                payment.token_nonce,
+            );
+
+            current_token_data.amount = payment.amount.clone();
+
+            if self.is_sovereign_chain().get() {
+                self.send().esdt_local_burn(
                     &payment.token_identifier,
                     payment.token_nonce,
+                    &payment.amount,
                 );
-                all_token_data.push(current_token_data.into());
-            } else {
-                all_token_data.push(StolenFromFrameworkEsdtTokenData::default());
-            }
-            
-            event_payments.push(
-                MultiValue3 ((
+
+                event_payments.push(MultiValue3((
                     payment.token_identifier.clone(),
                     payment.token_nonce,
-                    payment.amount.clone()    
-                ))
-            );
+                    current_token_data.clone(),
+                )));
+            } else {
+                let sov_token_id = self
+                    .multiversx_to_sovereign_token_id(&payment.token_identifier)
+                    .get();
+
+                if !sov_token_id.is_valid_esdt_identifier() {
+                    event_payments.push(MultiValue3((
+                        payment.token_identifier,
+                        payment.token_nonce,
+                        current_token_data.clone(),
+                    )));
+
+                    continue;
+                }
 
-            if burn_mapper.contains(&payment.token_identifier) {
                 self.send().esdt_local_burn(
                     &payment.token_identifier,
                     payment.token_nonce,
                     &payment.amount,
                 );
+
+                let mut sov_token_nonce = 0;
+
+                if payment.token_nonce > 0 {
+                    sov_token_nonce = self
+                        .multiversx_esdt_token_info_mapper(
+                            &payment.token_identifier,
+                            &payment.token_nonce,
+                        )
+                        .take()
+                        .token_nonce;
+
+                    self.sovereign_esdt_token_info_mapper(&sov_token_id, &sov_token_nonce)
+                        .take();
+                }
+
+                event_payments.push(MultiValue3((
+                    sov_token_id,
+                    sov_token_nonce,
+                    current_token_data.clone(),
+                )));
             }
```

Comment:
Maybe you can extract some (or all) of this logic into some separate functions. Try to keep endpoint bodies smaller

Reply:
I refactored the deposit endpoint